### PR TITLE
feat: return eoa deployment fee to deployer

### LINKED ISSUE #668: feat: update EOA management, so that a deployer of an EOA account can recover the fees spent on it
# Feature Request

Update EOA deployment management to allow a deployer of an account to recover the fee spent for deploying this account.

We would need a storage variable for this, that can be updated periodically later on to keep track of latest account deployment fees. 

Then, inside `[deploy_externally_owned_account](https://github.com/kkrt-labs/kakarot/blob/03f3686780f6b0145b575de3168e7ebd1c18593f/src/kakarot/library.cairo#L282)` the fees can be paid back to the caller of the function.
```[tasklist]
### Tasks
- [x] update EOA management to so that deployer can recover fee
- [x] add tests for the same
```

### PR DESCRIPTION
**Pull request by bajpai244**
Improves EOA deployment management, with this any Starknet account should be able to deploy a Kakarot Account, and they should be able to redeem back the fees spent for this operation.

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Currently if someone deploys an account via [Kakarot.deploy_externally_owned_account](https://github.com/kkrt-labs/kakarot/blob/03f3686780f6b0145b575de3168e7ebd1c18593f/src/kakarot/library.cairo#L282), they would have to spend fees for that and they won't be able to redeem it.

Resolves #668

## What is the new behavior?

- Anyone can deploy a Kakarot Account for an eoa
- Address needs to have a certain amount of eth already at their address to get a deployment
- The deployer can recover the amount spent for this deployment

### CODE REVIEW

File: scripts/deploy_kakarot.py [hunk 1]
```diff
@@ -2,7 +2,12 @@
 import logging
 from asyncio import run
 
-from scripts.constants import COMPILED_CONTRACTS, ETH_TOKEN_ADDRESS, EVM_ADDRESS
+from scripts.constants import (
+    COMPILED_CONTRACTS,
+    ETH_TOKEN_ADDRESS,
+    DEPLOY_FEE,
```
Comment (ClementWalter) [inline scripts/deploy_kakarot.py:8]: run `make format` I guess that this doesn't pass

File: src/kakarot/library.cairo [hunk 2]
```diff
@@ -192,6 +195,27 @@ namespace Kakarot {
         return (native_token_address_,);
     }
 
+     // @notice Set the deploy fee for deploying EOA on Kakarot.
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:198]: format

File: scripts/deploy_kakarot.py (hunk 1 above)
Comment (bajpai244) [reply scripts/deploy_kakarot.py:7]: done!

File: src/kakarot/library.cairo [hunk 3]
```diff
@@ -285,8 +309,23 @@ namespace Kakarot {
...
+        let (caller_address) = get_caller_address();
         let (class_hash) = externally_owned_account_class_hash.read();
         let (starknet_contract_address) = Accounts.create(class_hash, evm_contract_address);
+
+        let (local native_token_address) = get_native_token();
+        let (local deploy_fee) = get_deploy_fee();
+
+        let amount = Helpers.to_uint256(deploy_fee);
+        let (success) = IERC20.transferFrom(
+            contract_address=native_token_address,
```
Comment (ClementWalter) [inline src/kakarot/library.cairo:323]: format

File: src/kakarot/library.cairo (hunk 2 above)
Comment (bajpai244) [reply src/kakarot/library.cairo]: done!

File: src/kakarot/kakarot.cairo [hunk 4]
```diff
@@ -71,6 +73,26 @@ func get_native_token{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_chec
     return Kakarot.get_native_token();
 }
 
+// @notice Set the deploy fee for deploying EOA on Kakarot.
```
Comment (ClementWalter) [inline src/kakarot/kakarot.cairo:76]: format

File: scripts/utils/kakarot.py [hunk 5]
```diff
@@ -228,6 +228,9 @@ async def deploy_and_fund_evm_address(evm_address: str, amount: float):
     """
     Deploy an EOA linked to the given EVM address and fund it with amount ETH
     """
+
```
Comment (bajpai244) [inline scripts/utils/kakarot.py:231]: We are funding the account before, so that it has eth to pay back to the deployer.

File: tests/fixtures/ERC20.cairo [hunk 6]
```diff
@@ -6,7 +6,7 @@
 from starkware.cairo.common.cairo_builtins import HashBuiltin
 from starkware.cairo.common.uint256 import Uint256
 
-from openzeppelin.token.erc20.library import ERC20
+from tests.fixtures.lib.erc20 import ERC20
```
Comment (ClementWalter) [inline tests/fixtures/ERC20.cairo:9]: why this change?

File: src/kakarot/library.cairo (hunk 3 above)
Comment (bajpai244) [reply src/kakarot/library.cairo:323]: done [here](https://github.com/kkrt-labs/kakarot/pull/669/commits/e6c6e89d056885547c5df8e2f357d197bbfd944c).

File: tests/utils/accounts.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
+from starkware.starknet.testing.starknet import StarknetContract
+
+
+async def fund_evm_address(
```
Comment (ClementWalter) [inline tests/utils/accounts.py:4]: make it a fixture to have `kakarot` and `eth` already existing for the "end user" (the dev writing a test)

File: tests/fixtures/ERC20.cairo (hunk 6 above)
Comment (bajpai244) [reply tests/fixtures/ERC20.cairo:9]: the openzepplin contracts have an [assert](https://github.com/OpenZeppelin/cairo-contracts/blob/4dd04250c55ae8a5bbcb72663c989bb204e8d998/src/openzeppelin/token/erc20/library.cairo#L188), which reverts if the transfer is being done to the 0 address, in our test runner, what I can see is regardless of whatever I send the caller_address, inside the test it was 0, hence this was reverting, so I have changed the library to disable it.

File: tests/utils/constants.py [hunk 8]
```diff
@@ -1,5 +1,7 @@
 CHAIN_ID = 1263227476  # KKRT (0x4b4b5254) in ASCII
 
+DEPLOY_FEE = 10000
```
Comment (bajpai244) [inline tests/utils/constants.py:3]: I am not importing the constant from scripts here because if I do so it will ask for STARKNET_NETWORK environment variable.

File: src/kakarot/kakarot.cairo (hunk 4 above)
Comment (bajpai244) [reply src/kakarot/kakarot.cairo:76]: done.

File: tests/integration/accounts/test_contract_account.py [hunk 9]
```diff
@@ -91,3 +94,19 @@ async def test_should_give_infinite_allowance_to_kakarot(
...
+        async def test_should_take_deployment_fees(self, kakarot, eth):
+            evm_address = int(generate_random_evm_address(), 16)
+            computed_starknet_address = (
+                await kakarot.compute_starknet_address(evm_address).call()
+            ).result[0]
+
+            amount = 100000
+            await fund_evm_address(evm_address, kakarot, eth, amount)
+
+            await kakarot.deploy_externally_owned_account(evm_address).execute()
```
Comment (ClementWalter) [inline tests/integration/accounts/test_contract_account.py:107]: could we use a non trivial `caller_address` so as to make sure that funds are not just sent to address 0

**Review by ClementWalter — CHANGES_REQUESTED** [review]
plz review in github yourself first

otherwise, some python modif and should be good

File: tests/integration/accounts/test_contract_account.py (hunk 9 above)
Comment (bajpai244) [reply tests/integration/accounts/test_contract_account.py:107]: It makes no difference from what I could see, the caller address is 0 regardless of whatever caller address you choose.

Comment (ClementWalter) [reply tests/integration/accounts/test_contract_account.py:107]: there is plenty of example, even in this file, where we use `.execute(caller_address=xxx)` that wouldn't pass otherwise so I think it's working. can you double check?

File: tests/fixtures/ERC20.cairo (hunk 6 above)
Comment (ClementWalter) [reply tests/fixtures/ERC20.cairo:9]: why would we mint to address `0`? I think that it's not necessary

File: tests/integration/accounts/test_contract_account.py (hunk 9 above)
Comment (bajpai244) [reply tests/integration/accounts/test_contract_account.py:106]: Yeah, I have tried it, the reason how I explored this issue was when I started seeing the error, cannot transfer to 0 address from the ERC20 library!

I tried changing the value multiple times, but it looked like the caller address from tx.info was always 0! But, I am fixing few tests, and lemme recheck again and get back!

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py [hunk 10]
```diff
@@ -1,38 +1,32 @@
 import pytest
+from tests.utils.constants import ACCOUNT_BALANCE
 
 
 @pytest.mark.asyncio
 @pytest.mark.Counter
```
Comment (ClementWalter) [inline tests/integration/solidity_contracts/PlainOpcodes/test_safe.py:6]: can you fix this copy/pasted mark also plz?

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py [hunk 11]
```diff
@@ -1,38 +1,32 @@
...
 class TestSafe:
     class TestReceive:
-        async def test_should_receive_eth(self, safe, owner, eth):
-            await eth.mint(owner.starknet_address, (int(1e9), 0)).execute(
-                caller_address=owner.starknet_address
-            )
-            await safe.deposit(value=int(1e9), caller_address=owner.starknet_address)
-            assert await safe.balance() == 1e9
+        async def test_should_receive_eth(self, safe, owner):
+            await safe.deposit(value= ACCOUNT_BALANCE, caller_address=owner.starknet_address)
```
Comment (ClementWalter) [inline tests/integration/solidity_contracts/PlainOpcodes/test_safe.py:11]: you definitely **need** to run `make format` (black here) on save

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py [hunk 12]
```diff
@@ -1,38 +1,39 @@
...
+
 
 @pytest.mark.asyncio
-@pytest.mark.Counter
+@pytest.mark.Safe
 @pytest.mark.usefixtures("starknet_snapshot")
 class TestSafe:
     class TestReceive:
-        async def test_should_receive_eth(self, safe, owner, eth):
-            await eth.mint(owner.starknet_address, (int(1e9), 0)).execute(
```
Comment (bajpai244) [inline tests/integration/solidity_contracts/PlainOpcodes/test_safe.py:10]: no need to mint here, as the account is already funded with (1e9) amount of tokens so that it could pay for its deployment fee.

Hence, the balance of token for `owner` is `1e9 - DEPLOY_FEE`,  which is represented by the constant `ACCOUNT_BALANCE` { this constant is later on used in other places }

File: tests/fixtures/ERC20.cairo (hunk 6 above)
Comment (bajpai244) [reply tests/fixtures/ERC20.cairo]: resolved, I think probably some human error on my end because of which I got confused.

File: tests/utils/accounts.py (hunk 7 above)
Comment (bajpai244) [reply tests/utils/accounts.py]: done.

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py (hunk 10 above)
Comment (bajpai244) [reply tests/integration/solidity_contracts/PlainOpcodes/test_safe.py]: done.

File: tests/integration/solidity_contracts/PlainOpcodes/test_safe.py (hunk 11 above)
Comment (bajpai244) [reply tests/integration/solidity_contracts/PlainOpcodes/test_safe.py]: gotcha!

Might be we can also add git hooks just like how we have at the RPC? Lemme know, I can create an issue for it.

File: tests/integration/accounts/test_contract_account.py [hunk 13]
```diff
@@ -91,3 +93,25 @@ async def test_should_give_infinite_allowance_to_kakarot(
...
+            await kakarot.deploy_externally_owned_account(evm_address).execute(
+                caller_address=4
+            )
+
+            # asserting that the balance of the account is the amount minus the deployment fee
+            assert (
+                await eth.balanceOf(computed_starknet_address).call()
+            ).result.balance.low == ACCOUNT_BALANCE
+
+            # asserting that the deployer gets the Deployment Fee back { initial balance of the deployer is 0 }
```
Comment (ClementWalter) [inline tests/integration/accounts/test_contract_account.py:116]: you should compare balance of deployer before and after and not assume that it was 0

**Review by ClementWalter — COMMENTED** [review]
why did you have to put different caller_address everywhere?

**Comment by bajpai244** [comment]
> why did you have to put different caller_address everywhere?

Initially it was to make sure that there is no other caller_address=4, because I make a check on balance of this address in a test after it does a deployment of an EOA.

So, I made caller_address=5, then I thought that for the other two caller_addresses as well I can make them 6 and 7 so that if someone writes a test in future for caller_address=5  just like I did for 4, they won't have to count the number of deployments 5 has made.

But, I think might be I have over optimized here, I also think I can make 5,6,7 and 2, so pushing with these changes.

File: tests/integration/accounts/test_contract_account.py (hunk 13 above)
Comment (bajpai244) [reply tests/integration/accounts/test_contract_account.py:118]: Done.

File: tests/integration/accounts/test_contract_account.py [hunk 14]
```diff
@@ -91,3 +93,27 @@ async def test_should_give_infinite_allowance_to_kakarot(
...
+            self, kakarot, eth, fund_evm_address
+        ):
+            evm_address = int(generate_random_evm_address(), 16)
+            computed_starknet_address = (
+                await kakarot.compute_starknet_address(evm_address).call()
+            ).result[0]
+
+            await fund_evm_address(evm_address)
+
+            assert (await eth.balanceOf(4).call()).result.balance.low == 0
```
Comment (ClementWalter) [inline tests/integration/accounts/test_contract_account.py:107]: please assign magic number to a variable before usage, and use the variable instead

File: tests/integration/accounts/test_contract_account.py [hunk 15]
```diff
@@ -91,3 +93,27 @@ async def test_should_give_infinite_allowance_to_kakarot(
                 )
                 == "Uint256(low=340282366920938463463374607431768211455, high=340282366920938463463374607431768211455)"
             )
+
+        async def test_should_take_deployment_fees(
```
Comment (ClementWalter) [inline tests/integration/accounts/test_contract_account.py:97]: forgot to say but this test is ill placed. This tests a kakarot entrypoint and should consequently be in the test_kakarot.py file, in a `TestDeployExternallyOwnedAccount` class

**Review by ClementWalter — COMMENTED** [review]
I'd like us to clean these `caller_address`, plus move the test to the `test_kakarot.py` file, and eventually add some other tests for this `deploy_externally_owned_account` function

**Comment by bajpai244** [comment]
> I'd like us to clean these `caller_address`, plus move the test to the `test_kakarot.py` file, and eventually add some other tests for this `deploy_externally_owned_account` function

- I have moved the test to `test_kakarot.py`
- I have create a fixture `deployer_address`, which is 2, and have used it in all places where I was using `caller_address=2`
- I am leaving the hardcoded value `caller_address=4` as it is, because it is only used in 1 test, and is supposed to be unique and not shared with other tests, hence not making a fixture for it now, but if need be to extend testing with this fixture we can make one in future for it.

File: tests/integration/accounts/test_contract_account.py (hunk 15 above)
Comment (bajpai244) [reply tests/integration/accounts/test_contract_account.py]: Can you explain the difference between these two tests:
- `def test_should_deploy_starknet_contract_at_corresponding_address`
- `def test_should_deploy_starknet_contract_with_correct_implementation`

File: tests/utils/helpers.py [hunk 16]
```diff
@@ -145,11 +145,6 @@ def generate_random_private_key(seed=0):
     return keys.PrivateKey(int.to_bytes(random.getrandbits(256), 32, "big"))
 
 
-def generate_random_evm_address(seed=0):
```
Comment (bajpai244) [inline tests/utils/helpers.py:148]: It has now been made a fixutre, you can check #671 for more context.

File: tests/integration/accounts/test_contract_account.py (hunk 14 above)
Comment (bajpai244) [reply tests/integration/accounts/test_contract_account.py]: resolved.

File: tests/integration/accounts/test_contract_account.py (hunk 15 above)
Comment (ClementWalter) [reply tests/integration/accounts/test_contract_account.py]: I was thinking about the class hash, to make sure that it actually deploys an EOA. But let' not block for this

**Review by ClementWalter — APPROVED** [review]
